gammapy/analysis/steps.py

```diff
@@ -0,0 +1,498 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""classes containing the analysis steps supported by the high level interface"""
+
+import abc
+import os
+from astropy.coordinates import SkyCoord
+from astropy.table import Table
+from regions import CircleSkyRegion
+from gammapy.data import DataStore
+from gammapy.datasets import Datasets, FluxPointsDataset, MapDataset, SpectrumDataset
+from gammapy.estimators import (
+    ExcessMapEstimator,
+    FluxPointsEstimator,
+    LightCurveEstimator,
+)
+from gammapy.makers import (
+    FoVBackgroundMaker,
+    MapDatasetMaker,
+    ReflectedRegionsBackgroundMaker,
+    RingBackgroundMaker,
+    SafeMaskMaker,
+    SpectrumDatasetMaker,
+    DatasetsMaker
+)
+from gammapy.maps import Map, MapAxis, RegionGeom, WcsGeom
+from gammapy.utils.pbar import progress_bar
+from gammapy.utils.scripts import make_path, make_name
+
+
+class AnalysisStepBase(abc.ABC):
+    tag = "analysis-step"
+    requires_datasets = False
+    requires_models = False
+
+    def __init__(self, analysis, name=None, overwrite=True):
```

Comment:
On a first look taking the analysis class on `__init__` does not make sense to me. This introduces a "cross dependency", while in fact I think it can be hierarchical in the sense that the `Analysis` class is built from `AnalysisStep` classes. It think this can be resolved by slightly refactoring the API of the `AnalysisStep` class, along the lines of:

```python3

class AnalysisStep:
    """Analysis step class"""
    tag = "analysis-step"
    def __init__(self, analysis_sub_config, overwrite=True, log=None):
        self.config = analysis_sub_config
        self.overwrite = overwrite
        
        if log is None:
            log = logging.getLogger(__name__)
        
        self.log = log
    
    @property
    def maker_config(self):
        # translate analysis sub config to Gammapy API config here
        return config
    
    def run(self, datasets, models=None):
        maker = Maker(**self.maker_config)
        # returning might be optional...changes could happen in place
        return datasets
    
    


```

Reply:
Maybe a minima, you could take the `AnalysisConfig` on init and pass the `Analysis` to run().

Reply:
for now changed to take `AnalysisConfig` on init and pass the `Analysis` to run() but in the next PR I will introduce a specific `AnalysisProducts` container to return outputs and pass data references on run().